### PR TITLE
MAINT: vendor `pocketfft` as git submodule

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
https://github.com/scipy/scipy/pull/15321#pullrequestreview-1822131264

#### What does this implement/fix?
Hey @peterbell10 , here is the only file with meaningful changes if we are to vendor https://github.com/mreineck/pocketfft as a git submodule. Tests pass for me locally, so if this looks good to you I'll go ahead and update the PR to vendoring as a git submodule.

#### Additional information
cc @mreineck

### CODE REVIEW

File: scipy/fft/_pocketfft/pocketfft_hdronly.h [hunk 1]
```diff
@@ -149,14 +149,14 @@ template<> struct VLEN<double> { static constexpr size_t val=2; };
...
 
-// MSVC does not provide the standard aligned allocation functions,
-// yet _still_ sets the feature test macro (__cpp_aligned_new)...
-// While there are replacement functions (_aligned_malloc & _aligned_free),
-// those produce crashes when used here, see GH-19726
-#if __cplusplus >= 201703L && !defined(_MSC_VER)
+// the __MINGW32__ part in the conditional below works around the problem that
+// the standard C++ library on Windows does not provide aligned_alloc() even
+// though the MinGW compiler and MSVC may advertise C++17 compliance.
+#if (__cplusplus >= 201703L) && (!defined(__MINGW32__)) && (!defined(_MSC_VER))
```
Comment (lucascolley) [inline scipy/fft/_pocketfft/pocketfft_hdronly.h:155]: gh-19761. Can this go upstream?

File: scipy/fft/_pocketfft/pocketfft_hdronly.h [hunk 2]
```diff
@@ -48,16 +48,16 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endif
 
 #ifndef POCKETFFT_CACHE_SIZE
-#define POCKETFFT_CACHE_SIZE 16
+#define POCKETFFT_CACHE_SIZE 0
```
Comment (peterbell10) [inline scipy/fft/_pocketfft/pocketfft_hdronly.h:51]: Looks like you'll need to define this variable in the build system.

**Comment by lucascolley** [comment]
To-do list after vendoring as git submodule:

- define `POCKETFFT_CACHE_SIZE 16` in the build system.

anything else or is that it @peterbell10 ?

**Comment by peterbell10** [comment]
The `aligned_alloc` change seems fine, and the remaining changes look superficial so this seems to be the only change necessary.

**Comment by lucascolley** [comment]
If CI goes green then I think this is ready (although someone should carefully check that my Meson changes are correct and working).

File: scipy/fft/_pocketfft/meson.build [hunk 3]
```diff
@@ -23,12 +23,16 @@ else
     pocketfft_threads += ['-DPOCKETFFT_PTHREADS']
   endif
 endif
+pocketfft_args = pocketfft_threads + ['-DPOCKETFFT_CACHE_SIZE=16']
+
+incdir = include_directories('../../_lib/pocketfft')
 
 py3.extension_module('pypocketfft',
   'pypocketfft.cxx',
   cpp_args: pocketfft_threads,
```
Comment (lucascolley) [inline scipy/fft/_pocketfft/meson.build:32]:
```suggestion
  cpp_args: pocketfft_args,
```

**Comment by rgommers** [comment]
The build config changes look fine to me; the one thing to add would be a guard for the submodule missing at the top of `scipy/_lib/meson.build`.

**Comment by lucascolley** [comment]
> The build config changes look fine to me; the one thing to add would be a guard for the submodule missing at the top of scipy/_lib/meson.build.

Done 👍

**Review by peterbell10 — APPROVED** [review]

**Review by tylerjereddy — APPROVED** [review]
Looks like comments from Ralf and Peter were addressed, and there's an approval from Peter.

The diff looks good to me as well, and I correctly got the right error message on this branch when I intentionally didn't update the submodule.

In it goes I think, good to let this sit in `main` for a little bit before the earlier-than-usual `1.13.0` gets branched.

**Comment by tylerjereddy** [comment]
thanks

**Comment by rgommers** [comment]
Thanks Lucas & reviewers!

Just to make 100% sure: @mreineck is your mirror on GitHub guaranteed to be stable and not disappear in the future? Such an event would make checking out older versions of SciPy very hard, so if you'd prefer not to make a hard guarantee then we can also keep a fork of that repo under the scipy org (we have a few other git submodules there, like `unuran` and `PROPACK`).

**Comment by mreineck** [comment]
I certainly don't plan to remove it at any point! But on the other hand I don't mind at all if you want to play it extra safe and keep a fork on scipy.org ... and keeping things in sync should be easy, since it'll be bug fixes only.

**Comment by rgommers** [comment]
Thanks @mreineck. I think extra safe may be wise here, and it's low cost to do indeed. The risk is low here, but we're accumulating more and more git submodules - and multiple small risks may start to add up to something more likely to materialize eventually.

**Comment by rgommers** [comment]
Done in gh-19916.

**Pull request merged**